### DISCUSSION TITLE
How can I set the transductive environment manually?

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#9387] [Q&A]
Greetings, I have a question about the transductive setting (i.e., the nodes in the training set should be in the validation/test set) on the link prediction task.
In my case, I have positive and negative edge sets for the model training.
(Note that the negative edges are not the randomly selected unknown links.)
However, there were some unique nodes and they disturbed data splitting since they appeared only one time across the training-validation-test dataset.
To avoid this situation, I tried to replace some of the major nodes with the unique nodes but there were always unique genes at least one.

Can anyone give me feedback on my code?
```python
num_nodes = 50
num_edges = 100
row = torch.randint(0, num_nodes, (num_edges,))
col = torch.randint(0, num_nodes, (num_edges,))
edge_index = torch.stack([row, col], dim=0)
edge_index = edge_index[:, row < col] # make directed

print(edge_index)
#tensor([[23, 45, 24, 17, 30, 19,  5, 25, 16, 22,  6, 10, 14, 26,  9, 12, 20, 45,
#         16, 27,  3,  2,  2, 19, 26,  5,  6,  1, 13, 23,  8, 37,  4, 21,  9, 14,
#          5, 29, 14, 17, 11, 11, 18, 18],
#        [36, 48, 48, 38, 47, 30, 30, 36, 43, 29, 44, 15, 32, 31, 25, 37, 37, 46,
#         44, 42, 21, 27, 40, 30, 45, 32, 34,  5, 34, 47, 31, 48, 45, 43, 30, 45,
#         33, 47, 25, 23, 18, 49, 19, 46]])

n_id, g_counts = torch.unique(edge_index, return-counts=True)
unique_node = n_id[n_counts==1]
major_node = n_id[n_counts>2]

# The location of the index which has a major node in either source and target
major_s_idx = (edge_index[..., None] == major_node).any(-1)[0].nonzero()[:, 0]
major_t_idx = (edge_index[..., None] == major_node).any(-1)[1].nonzero()[:, 0]

#trial 1
replaced_edge_index = edge_index.clone()
replaced_edge_index[0][major_s_idx[torch.randint(0, len(major_s_idx), size=(len(unique_node), ))]] = unique_node
replaced_edge_index[1][major_t_idx[torch.randint(0, len(major_t_idx), size=(len(unique_node), ))]] = unique_node
print(replaced_edge_index)
#tensor([[33, 45, 24, 17, 15, 20,  8, 10, 16, 22,  6, 10, 42, 26,  9, 12, 20, 49,
#         16, 27,  3,  2,  2, 19, 26,  4,  6,  1, 13,  3,  8, 38,  4, 21,  9, 14,
#         12, 29, 14, 17, 11, 11, 24, 40],
#        [36, 48, 20, 38, 47, 30, 22, 36, 43, 29, 44, 15, 32, 31, 42, 49, 37, 46,
#         44, 42, 21, 27, 40, 30, 45, 32, 34,  5, 34, 40, 31, 48, 12, 43, 30, 10,
#         33, 47, 24,  4, 18, 49, 33, 46]])

#trial 2
replaced_edge_index = edge_index.clone()
replaced_edge_index[0][major_s_idx[torch.randperm(len(major_s_idx))[:len(unique_genes)]]] = unique_node
replaced_edge_index[1][major_t_idx[torch.randperm(len(major_t_idx))[:len(unique_genes)]]] = unique_node

n_id_re, n_counts_re = torch.unique(replaced_edge_index, return_counts=True)
print(n_id_re[n_counts_re==1])
# 6
```
